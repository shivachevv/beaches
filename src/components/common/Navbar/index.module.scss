
.menu {
    div:nth-child(3) {
        top: auto !important;
        bottom: 60px!important;
    }
}
.test {
    background-color: #b06d8d !important;
}

.link {
    text-decoration: none;
    color: inherit;
    text-transform: uppercase;
  }