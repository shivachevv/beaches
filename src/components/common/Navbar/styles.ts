const styles = {
  link: {
    textDecoration: "none",
    color: "inherit",
    textTransform: "uppercase",
  },
};

export default styles;
