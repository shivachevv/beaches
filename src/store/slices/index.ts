import { combineReducers, Reducer } from "redux";
import authReducer from "./auth";
import beachesReducer from "./beaches";
import reservationsReducer from "./reservations";
import { RootState } from "../../interfaces";

const rootReducer: Reducer<RootState> = combineReducers<RootState>({
  auth: authReducer,
  beaches: beachesReducer,
  reservations: reservationsReducer,
});

export default rootReducer;
