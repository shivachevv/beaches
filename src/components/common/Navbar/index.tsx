import * as React from "react";
import {
  AppBar,
  Avatar,
  Box,
  Button,
  Drawer,
  IconButton,
  Menu,
  MenuItem,
  Toolbar,
  Typography,
  useMediaQuery,
  useTheme,
} from "@mui/material";
import MenuIcon from "@mui/icons-material/Menu";
import { Link, useNavigate } from "react-router-dom";
import logo from "../../../assets/images/beaches_logo.png";
import styles from "./styles";
import { useStyles } from "../../../utils/helpers";
import { NavLinksContext } from "../../../contexts/NavLinksProvider";
import { useAppDispatch, useAppSelector } from "../../../store/hooks";
import { logout } from "../../../store/slices/auth";
import scss from "./index.module.scss";

type Props = Record<string, unknown>;

const Navbar: React.FC<Props> = (props: Props) => {
  const navigate = useNavigate();
  const navLinks = React.useContext(NavLinksContext);
  const { isAuthenticated, currentUser } = useAppSelector(
    (state) => state.auth
  );

  const dispatch = useAppDispatch();

  const logoutUser = () => {
    dispatch(logout());
  };

  const classes = useStyles(styles);
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down("sm"));
  const menuPosition = {
    top: isMobile ? "auto" : "0",
    bottom: isMobile ? "0" : "auto",
  };
  const [isMobileNavOpen, setIsMobileNavOpen] = React.useState<boolean>(false);
  const toggleMenu = ({ state }: { state: boolean }) => {
    setIsMobileNavOpen(state);
  };

  const renderNavLinks = ({ isMobile }: { isMobile: boolean }) =>
    navLinks.map((page) => (
      <Link to={page.path} key={page.path} className={classes.link}>
        <Button sx={{ my: 2, color: isMobile ? "" : "white" }}>
          {page.name}
        </Button>
      </Link>
    ));

  const getUserLetters = () =>
    `${currentUser?.firstName[0]} ${currentUser?.lastName[0]}`.toLocaleUpperCase();

  const [userMenuAnchor, setUserMenuAnchor] =
    React.useState<null | HTMLElement>(null);
  const openUserMenu = (event: React.MouseEvent<HTMLElement>) => {
    setUserMenuAnchor(event.currentTarget);
  };
  const closeUserMenu = () => {
    setUserMenuAnchor(null);
  };
  const renderUserDetails = (): React.ReactNode => {
    return isAuthenticated ? (
      <Box
        sx={{
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
        }}
      >
        <IconButton onClick={openUserMenu}>
          <Avatar sx={{ bgcolor: "orange" }}>{getUserLetters()}</Avatar>
        </IconButton>
        <Menu
          sx={{ mt: "45px" }}
          id="menu-appbar"
          anchorEl={userMenuAnchor}
          anchorOrigin={{
            vertical: "top",
            horizontal: "right",
          }}
          keepMounted
          transformOrigin={{
            vertical: "top",
            horizontal: "right",
          }}
          open={Boolean(userMenuAnchor)}
          onClose={closeUserMenu}
          className={isMobile ? scss.menu : ""}
        >
          <MenuItem onClick={closeUserMenu}>
            <Link to="/my-profile" className={classes.link}>
              My Profile
            </Link>
          </MenuItem>
          <MenuItem onClick={closeUserMenu}>
            <Button color="inherit" fullWidth onClick={logoutUser}>
              Logout
            </Button>
          </MenuItem>
        </Menu>
      </Box>
    ) : (
      <Button color="inherit" onClick={() => navigate("/login")}>
        Login
      </Button>
    );
  };

  return (
    <AppBar
      position="absolute"
      color="primary"
      sx={{ top: menuPosition.top, bottom: menuPosition.bottom }}
    >
      <Toolbar>
        <Typography>
          <Box
            component="img"
            sx={{
              height: 30,
              width: 30,
            }}
            alt="Logo"
            src={logo}
          />
        </Typography>
        <Box
          sx={{
            flexGrow: 1,
            display: {
              xs: "none",
              sm: "flex",
              justifyContent: "space-between",
            },
          }}
        >
          <Box>{renderNavLinks({ isMobile: false })}</Box>
          {!isMobile && renderUserDetails()}
        </Box>
        <Box
          sx={{
            flexGrow: 1,
            display: {
              xs: "flex",
              justifyContent: "space-between",
              sm: "none",
            },
          }}
        >
          <IconButton
            size="large"
            aria-controls="menu-appbar"
            aria-haspopup="true"
            onClick={() => toggleMenu({ state: true })}
            color="inherit"
          >
            <MenuIcon />
          </IconButton>
          {isMobile && renderUserDetails()}
          <Drawer
            anchor="bottom"
            open={isMobileNavOpen}
            onClose={() => toggleMenu({ state: false })}
          >
            {renderNavLinks({ isMobile: true })}
          </Drawer>
        </Box>
      </Toolbar>
    </AppBar>
  );
};

export default Navbar;
