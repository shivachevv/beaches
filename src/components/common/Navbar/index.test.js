/* eslint-disable no-undef */
import React from "react";
import Navbar from "./index";
import { render, cleanup } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../../../store";

afterEach(cleanup);

const renderedNav = (
  <Provider store={store}>
    <Router>
      <Navbar />
    </Router>
  </Provider>
);

it("should take a snapshot", () => {
  const { asFragment } = render(renderedNav);

  expect(asFragment(<Navbar />)).toMatchSnapshot();
});

it("should render user details", () => {
  const component = render(renderedNav);
  expect(component.getByTestId("user-details")).toHaveTextContent(
    "LoginRegister"
  );
});
