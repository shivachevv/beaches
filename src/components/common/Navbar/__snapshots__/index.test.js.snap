// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionAbsolute css-1vsh8fa-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        <img
          alt="Logo"
          class="MuiBox-root css-1nsz27u"
          src="beaches_logo.png"
        />
      </p>
      <div
        class="MuiBox-root css-19wiw76"
      >
        <div
          class="MuiBox-root css-n085mf"
        >
          <a
            class="makeStyles-link-1"
            href="/"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1jlq4mj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Home
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            class="makeStyles-link-1"
            href="/my-profile"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1jlq4mj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              My Profile
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
          <a
            class="makeStyles-link-1"
            href="/reservations/:userId"
          >
            <button
              class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1jlq4mj-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              My Reservations
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
        </div>
        <div
          class="MuiBox-root css-1n2mv2k"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1y942vo-MuiButtonBase-root-MuiButton-root"
            data-testid="user-details-login"
            tabindex="0"
            type="button"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit MuiButtonBase-root  css-1gj1zu1-MuiButtonBase-root-MuiButton-root"
            data-testid="user-details-register"
            tabindex="0"
            type="button"
          >
            Register
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiBox-root css-14ez9k8"
      >
        <button
          aria-controls="menu-appbar"
          aria-haspopup="true"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
